blog/source/_posts/MIDI-velocity-transformer.md

```diff
@@ -0,0 +1,266 @@
+---
+title: MIDI Velocity Prediction with Transformer
+date: 2023-09-11
+tags:
+author: wmatejuk
+---
+
+### Why is it important?
+
+{% algrtmImgBanner MIDI-velocity-transformer/banner-1.jpg pianoroll %}
+
+MIDI velocity is a crucial element in music dynamics, determining the force with which a note is played, 
+which profoundly influences the emotional quality of music. 
+
+If you were to take a sequence of notes and predict their velocities by an untrained model, this is what you wold have ended up with:
+<div style="-webkit-column-count: 2; -moz-column-count: 2; column-count: 2;">
+    
+  **Original** 
+        
+  **Predicted by untrained model**
+
+</div>
+
+
+<div style="-webkit-column-count: 2; -moz-column-count: 2; column-count: 2;">
+
+  {% algrtmImg MIDI-velocity-transformer/samples/9115-real.png pianoroll 170px %}
+  {% algrtmAudio MIDI-velocity-transformer/samples/9115-real.mp3 %}
+        
+  {% algrtmImg MIDI-velocity-transformer/samples/9115-pred-untrained.png pianoroll 170px %}
+  {% algrtmAudio MIDI-velocity-transformer/samples/9115-pred-untrained.mp3 %}
+
+</div>
+
+
+Sounds pretty bad, doesn't it?
+
+Our Transformer-based model aims to decode 
+this nuanced aspect of musical expression, unraveling the hidden patterns 
+within quantized MIDI data.
+
+
+### Model Overview
+    
+*The Transformer model* is ideal for this task because it excels at capturing complex dependencies in sequential data, making it well-suited for predicting MIDI velocities accurately.
+
+This model's suitability arises from its *self-attention* mechanism,
+which enables it to weigh the importance of different parts of the input sequence,
+regardless of their temporal order.
+In the context of MIDI data, this means that the Transformer can effectively learn
+and leverage complex relationships between musical notes, their timing,
+and how these factors influence the resulting velocity.
+
+Strong prediction results signify the model's proficiency in extracting vital
+features and comprehending intricate relationships. 
+Its accurate encoding of quantized MIDI data and precise velocity predictions
+mark a significant stride toward the realm of emotionally resonant AI music generation.
+### Data Preprocessing
+#### MIDI data
+MIDI data describes notes by 5 features:
+   1. Pitch - Represented as a number between 0 and 127 (or 21 to 108 for piano keys, reflecting the standard 88-key keyboard).
+   2. Start - Indicates the moment a key is pressed, measured in seconds.
+   3. End - Marks the second when the key is released.
+   4. Duration - calculated as the time elapsed between the key's press and release.
+   5. Velocity - ranging from 0 to 128, indicating the intensity of the key press.
```

Comment:
List of points is not the best format to describe a data structure. One alternative would be to make a code snippet showing a "note" class design, and referring to piano performance as a list of notes:

```python
class Note:
    pitch: int
    start: float
    end: float
    velocity: int
```

You can also use https://mermaid.live/ to make a class diagram, I think our blog should support it just like github:

```mermaid
classDiagram
    class Note{
      pitch: int
      velocity: int
      start: float
      end: float
    }
```

(this does not look great, but you could play around with it)

Reply:
I do not think it does support mermaid :(

Reply:
I just updated `master`, and now it does 🎉

Reply:
:fire::fire::fire: